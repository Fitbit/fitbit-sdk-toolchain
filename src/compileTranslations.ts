import { basename } from 'path';
import { Transform } from 'stream';

import { TranslationLoader } from '@fitbit/bison-i18n';
import humanizeList from 'humanize-list';
import PluginError from 'plugin-error';
import Vinyl from 'vinyl';

import { validateLanguageTag } from './languageTag';
import { Locales } from './ProjectConfiguration';

const PLUGIN_NAME = 'compileTranslations';

/**
 * Compile .po files into Bison FW binary translation files.
 *
 * The emitted files have a .translationLanguage property which holds
 * the string BCP 47 language tag for the translations in the file.
 */
export default function compileTranslations(defaultLanguage: string) {
  const translations = new TranslationLoader();
  const translationFiles = new Map<string, Vinyl>();

  const transform = new Transform({
    objectMode: true,
    transform(file: Vinyl, _, next) {
      if (file.isNull() || file.extname !== '.po') {
        return next(undefined, file);
      }

      const languageTag = basename(file.basename, '.po');
      if (!validateLanguageTag(languageTag)) {
        next(
          new PluginError(
            PLUGIN_NAME,
            // tslint:disable-next-line:max-line-length
            `Translation file ${
              file.basename
            } has a bad name. Translation files must be named ${humanizeList(
              Object.keys(Locales).map((tag) => tag + '.po'),
              { conjunction: 'or' },
            )}.`,
            { fileName: file.relative },
          ),
        );
        return;
      }

      if (file.isBuffer()) {
        try {
          const contents = file.contents.toString('utf-8');
          translations.loadLanguage(languageTag, contents);
          translationFiles.set(languageTag, file.clone({ contents: false }));
          return next();
        } catch (error) {
          next(
            new PluginError(PLUGIN_NAME, error as Error, {
              fileName: file.relative,
            }),
          );
        }
      } else {
        next(
          new PluginError(
            PLUGIN_NAME,
            file.isStream()
              ? 'Streaming mode is not supported.'
              : 'Internal error processing file.',
            { fileName: file.relative },
          ),
        );
        return;
      }
    },

    flush(done) {
      const languageTable = translations.build();

      if (
        languageTable.languages.size > 0 &&
        !languageTable.languages.has(defaultLanguage)
      ) {
        done(
          new PluginError(
            PLUGIN_NAME,
            `No translation file found for default language "${defaultLanguage}"`,
          ),
        );
        return;
      }

      for (const language of languageTable.languages) {
        const file = translationFiles.get(language)!;

        try {
          file.contents = Buffer.from(languageTable.getLanguage(language));
          file.translationLanguage = language;
        } catch (error) {
          done(
            new PluginError(PLUGIN_NAME, error as Error, {
              fileName: file.relative,
            }),
          );
          return;
        }

        // Not a constant because the actual output path is arbitrary and
        // can be changed at any time without impact. The path is kept
        // short because of limitations on the max size of manifest.json.
        file.path = `l/${language}`;
        transform.push(file);
      }

      done();
    },
  });

  return transform;
}
