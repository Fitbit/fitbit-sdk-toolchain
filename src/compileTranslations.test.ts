import Vinyl from 'vinyl';

import compileTranslations from './compileTranslations';
import { PassThrough, Writable } from 'stream';

const translationContents = 'msgid "foo"\nmsgstr "foo"';

function expectTranslationsForLanguages(
  done: jest.DoneCallback,
  ...languages: string[]
) {
  const expected = new Set(languages);
  const actual = new Set<string>();

  return new Writable({
    objectMode: true,
    write(file: Vinyl, _, next) {
      if (actual.has(file.translationLanguage)) {
        done.fail(`Seen translations for ${file.translationLanguage} twice`);
      }
      actual.add(file.translationLanguage);
      next();
    },

    final(callback) {
      expect(actual).toEqual(expected);
      callback();
      done();
    },
  });
}

function translationFileSource(...paths: string[]) {
  const stream = new PassThrough({ objectMode: true });

  paths.forEach((path) =>
    stream.write(
      new Vinyl({
        path,
        contents: Buffer.from(translationContents),
      }),
    ),
  );

  stream.end();

  return stream;
}

describe('when no files are piped', () => {
  it('succeeds with empty output', (done) => {
    compileTranslations('en')
      .on('error', done.fail)
      .resume()
      .on('end', done)
      .end();
  });
});

describe('when no .po files are piped in', () => {
  it('passes the files through', (done) => {
    const file = new Vinyl({
      path: 'en.po.txt',
      contents: Buffer.from('foo'),
    });

    const files = new Map<string, Vinyl>();

    compileTranslations('en')
      .on('error', done.fail)
      .on('data', (file: Vinyl) => files.set(file.relative, file))
      .on('end', () => {
        expect(files).toEqual(new Map<string, Vinyl>([['en.po.txt', file]]));
        done();
      })
      .end(file);
  });
});

it('grabs .po files from any directory', (done) => {
  translationFileSource('en-US.po', 'foo/es-ES.po', 'foo/bar/fr-FR.po')
    .pipe(compileTranslations('en-US'))
    .on('error', done.fail)
    .pipe(expectTranslationsForLanguages(done, 'en-US', 'es-ES', 'fr-FR'));
});

function compileExpectError(defaultLanguage: string, done: jest.DoneCallback) {
  return compileTranslations(defaultLanguage)
    .on('data', () => done.fail('got unexpected data'))
    .on('error', (error) => {
      expect(error).toMatchSnapshot();
      done();
    });
}

describe('rejects .po files whose names are not acceptable language tags', () => {
  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>(['e.po', 'en-USA.po', 'sl-IT-nedis.po'])(
    '%s',
    (path: string, done: jest.DoneCallback) => {
      compileExpectError('en', done).end(
        new Vinyl({
          path,
          contents: Buffer.from('foo'),
        }),
      );
    },
  );
});

it('bails when multiple .po files of the same name are present', (done) => {
  translationFileSource('en-US.po', 'a/en-US.po').pipe(
    compileExpectError('en-US', done),
  );
});

it('bails when the default langugage is not present', (done) => {
  translationFileSource('en-US.po').pipe(compileExpectError('es-ES', done));
});

it('bails when a .po file is malformed', (done) => {
  compileExpectError('en-US', done).end(
    new Vinyl({
      path: 'en-US.po',
      contents: Buffer.from('definitely not gettext format'),
    }),
  );
});

it('gracefully fails when a streaming-mode file is passed in', (done) => {
  compileExpectError('en-US', done).end(
    new Vinyl({
      path: 'en-US.po',
      contents: new PassThrough(),
    }),
  );
});

it('renames the transformed file, replacing the whole relative path', (done) => {
  translationFileSource('a/b/c/en-US.po')
    .pipe(compileTranslations('en-US'))
    .on('data', (file: Vinyl) => {
      expect(file.relative).toBe('l/en-US');
      done();
    })
    .on('error', done.fail);
});
