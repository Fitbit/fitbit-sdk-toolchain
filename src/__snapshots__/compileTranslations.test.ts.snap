// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bails when a .po file is malformed 1`] = `
PluginError {
  "fileName": "en.po",
  "message": "There are no translation strings for \\"en\\"",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`bails when multiple .po files of the same name are present 1`] = `
PluginError {
  "fileName": "a/en.po",
  "message": "Language 'en' already loaded",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`bails when the default langugage is not present 1`] = `
PluginError {
  "message": "No translation file found for default language \\"es\\"",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`gracefully fails when a streaming-mode file is passed in 1`] = `
PluginError {
  "fileName": "en.po",
  "message": "Streaming mode is not supported.",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`rejects .po files whose names are not acceptable language tags e.po 1`] = `
PluginError {
  "fileName": "e.po",
  "message": "Translation file e.po has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`rejects .po files whose names are not acceptable language tags en-USA.po 1`] = `
PluginError {
  "fileName": "en-USA.po",
  "message": "Translation file en-USA.po has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;

exports[`rejects .po files whose names are not acceptable language tags sl-IT-nedis.po 1`] = `
PluginError {
  "fileName": "sl-IT-nedis.po",
  "message": "Translation file sl-IT-nedis.po has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)",
  "name": "Error",
  "plugin": "compileTranslations",
}
`;
